### PR TITLE
Add test-conventions skill

### PR DESCRIPTION
**Pull request by jcdendrite**
## Summary
- Adds a new auto-triggered Claude Code skill (`user-invocable: false`) that provides language-agnostic testing conventions and principles
- Triggers when discussing test strategy, planning tests, reviewing coverage, or writing test infrastructure
- Covers: test pyramid (unit/integration/contract/smoke), design for testability, test isolation (DB, parallel execution), naming conventions, test data management, coverage judgment, mock design, flaky test diagnosis, and 16 anti-patterns
- Reviewed by Staff Backend Engineer and Senior SDET agents — all HIGH and MEDIUM findings addressed

## Key sections
1. **Test pyramid** — includes contract testing layer, unit test speed thresholds, and "signs of an inverted pyramid" diagnostics
2. **Design for testability** — DI, language-aware error handling advice, test double taxonomy (stub/mock/fake/spy)
3. **Test isolation** — global state, DB transaction rollback, parallel-safe test guidance
4. **Naming and structure** — scenario+outcome naming, Arrange/Act/Assert, regression intent comments
5. **Test data** — factory/builder patterns, magic value avoidance, shared DB collision prevention
6. **Coverage judgment** — split HTTP-level vs service-level integration tests, concurrency coverage, security allow+deny paths
7. **Mock design** — fidelity, mutation recording, tautological mock detection with concrete example
8. **Flaky tests** — expanded root-cause enumeration (TZ, locale, floats, async, iteration order), in-test vs CI-level retry distinction
9. **Anti-patterns** — 16 entries including assertion-free tests, sleep-based sync, over-mocking, test interdependence

## Test plan
- [ ] Verify skill triggers correctly in a test strategy conversation
- [ ] Verify skill does NOT trigger when a project-level test skill is loaded
- [ ] Verify skill does NOT trigger for mechanical tasks (running tests, fixing assertions)
- [ ] Review advice accuracy across Python, Go, TypeScript, and Java contexts
- [ ] Confirm 225 lines is within acceptable context budget (~3-5K tokens on trigger)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: claude/.claude/skills/test-conventions/SKILL.md [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+**Contract tests** verify that service-to-service API schemas stay compatible without requiring a running instance of the other service. Use them at any service boundary where teams deploy independently.
+
+### Signs of an inverted pyramid
+- Most tests call real services (DB, APIs) to test internal branching logic
+- Tests are slow (>1s each) because of network round-trips
+- Individual unit tests taking >50-100ms, indicating a hidden real dependency or overly complex setup
+- External API rate limits constrain how often you can run the test suite
+- Tests are flaky because of network timeouts or service availability
+- You can't run the test suite offline or in CI without credentials
+- Adding a new code branch requires setting up complex test state in a real database
```
Comment (jcdendrite) [inline claude/.claude/skills/test-conventions/SKILL.md]: This might be a useful skill when debugging why tests are slow or when evaluating a repo, but I don't think it's helpful to reference all the time when writing tests. Thoughts?

File: claude/.claude/skills/test-conventions/SKILL.md [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+**Integration → unit when:**
+- The test sets up complex DB state to exercise one `if` branch
+- The test hits an external API just to test error handling
+- The test is slow (>500ms) and tests pure logic, not wiring
+- You're hitting rate limits on external APIs
+
+**Smoke → integration when:**
+- The test calls a real third-party API to verify your code handles errors
+- The test is flaky because of network conditions
+- The test creates real resources (emails, contacts) as a side effect
```
Comment (jcdendrite) [inline claude/.claude/skills/test-conventions/SKILL.md]: This shouldn't be stated as an absolute, no-exceptions rule. This is more "when to consider moving a test down the pyramid".

File: claude/.claude/skills/test-conventions/SKILL.md [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+| **Fake** | Lightweight real implementation (in-memory DB, local HTTP server) | Testing without real infrastructure; also useful in unit tests when stubs are too complex (e.g., an in-memory repository that maintains state across calls) |
+| **Spy** | Wraps the real implementation, records calls | You need real behavior but want to verify interaction |
+
+Use the narrowest double that covers the test's intent. Prefer stubs for unit tests of pure logic; use mocks only when verifying interaction is the point of the test.
+
+### Test double seams by dependency type
+- **Database calls:** Stub/fake the client object, or use transaction rollback isolation (see section 3)
+- **External HTTP APIs:** Intercept by URL pattern or use a fake HTTP server
+- **Env vars / config:** Set and restore in setup/teardown blocks
+- **Time:** Inject timestamps as parameters, or set up data relative to "now"
```
Comment (jcdendrite) [inline claude/.claude/skills/test-conventions/SKILL.md]: Using data relative to "now" creates flaky tests. I say avoid unless absolutely necessary.

File: claude/.claude/skills/test-conventions/SKILL.md [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+
+## 3. Test isolation
+
+### Tests must be independent
+- Never rely on test execution order
+- Each test creates its own data and cleans up in teardown
+- Use unique identifiers (timestamps, counters) to prevent cross-test collision
+
+### Global state must be saved and restored
```
Comment (jcdendrite) [inline claude/.claude/skills/test-conventions/SKILL.md]: And avoid tests that require global state if there's an alternative that doesn't require it.

File: claude/.claude/skills/test-conventions/SKILL.md [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+When replacing globals (HTTP client, fetch function, clock), always restore the original in a guaranteed cleanup block, even if the test fails or throws.
+
+### Database test isolation
+- **Transaction rollback pattern:** Wrap each test in a transaction and roll back at the end. Standard in Django, Rails, Spring, and most ORMs — more efficient than truncation.
+- **Test database lifecycle:** Use a dedicated test database, run migrations before the suite, reset state between tests.
+- **In-memory vs. real engine:** In-memory databases (e.g., SQLite in-memory mode) are fast but have dialect differences (JSON columns, CTEs, locking behavior). When dialect fidelity matters, use the same engine as production.
+
+### Parallel-safe tests
+When tests run in parallel (pytest-xdist, Jest workers, Go's `t.Parallel()`, JUnit parallel mode):
+- Never bind to hardcoded ports; use port 0 or dynamic allocation
```
Comment (jcdendrite) [inline claude/.claude/skills/test-conventions/SKILL.md]: What about for running tests locally? Consider both local and CI testing.

File: claude/.claude/skills/test-conventions/SKILL.md [hunk 6]
```diff
@@ -0,0 +1,225 @@
...
+### What each test layer should verify
+
+**Unit tests:**
+- Every `if/else` branch in the function
+- Error return paths (missing input, API failure, invalid state)
+- Edge cases (zero values, null, empty strings, boundary conditions)
+- Side effects via mocks (did it call `.update()` with the right payload?)
+
+**Integration tests:**
+- **HTTP-level:** Auth rejection (401/403), authorized success (2xx), response shape (expected fields and types)
```
Comment (jcdendrite) [inline claude/.claude/skills/test-conventions/SKILL.md]: There's some overlap with contract tests when you talk about response shape. Should be more specific here.

File: claude/.claude/skills/test-conventions/SKILL.md [hunk 7]
```diff
@@ -0,0 +1,225 @@
...
+# Testing Conventions
+
+## 1. Test pyramid
+
+| Layer | Speed | External deps | What it covers | Volume |
+|-------|-------|---------------|----------------|--------|
+| **Unit** (stubbed deps) | Milliseconds | None | Branch logic, error paths, edge cases, data transformations | Many — every code path |
+| **Integration** (real local services) | Hundreds of ms | Local DB, local services | Auth boundaries, response contracts, wiring between layers | Few — happy path + key error paths per endpoint or boundary |
+| **Contract** (schema verification) | Milliseconds–seconds | None (consumer) or local (provider) | API schemas between services match consumer expectations | One per service boundary |
+| **Smoke / E2E** (real external APIs) | Seconds | Third-party APIs, real infra | Full flow works end-to-end | 1-2 per feature, run on deploy not on every commit |
```
Comment (jcdendrite) [inline claude/.claude/skills/test-conventions/SKILL.md]: Smoke tests and e2e tests are different entities for different purposes. They do involve actual real APIs, but I recommend not merging them.

File: claude/.claude/skills/test-conventions/SKILL.md [hunk 8]
```diff
@@ -0,0 +1,225 @@
...
+   - Locale-dependent formatting (number/date formats vary by system locale)
+   - Floating point comparison (use approximate equality, not exact)
+   - Async timing (asserting on async state without proper synchronization)
+   - Non-deterministic iteration order (hash maps/sets with no guaranteed order)
+   - Port or resource conflicts in parallel execution
+2. **Fix the layer** — if the test is flaky because it hits a real service to test logic, push it down the pyramid (see section 1)
+3. **Quarantine in-test retries** — retry loops inside test code mask the underlying issue. If a fix isn't immediate, quarantine the test (skip with a tracking issue) rather than letting it erode trust in the suite. CI-level retry policies (rerun failed tests once before failing the build) are a separate, reasonable practice for transient infrastructure issues — but track retry rates and investigate if a test needs retries frequently
+4. **Never delete without replacement** — a flaky integration test that covers an auth boundary still represents needed coverage. Replace it with a reliable test at the right layer before removing it
```
Comment (jcdendrite) [inline claude/.claude/skills/test-conventions/SKILL.md]: Might be good to put this in a test debugging or evaluation section or separate skill, like with the section about whether a pyramid is inverted. I'm not convinced that this list will actually help with writing and planning the tests in the first place.

File: claude/.claude/skills/test-conventions/SKILL.md [hunk 9]
```diff
@@ -0,0 +1,225 @@
...
+| Unconditional global state deletion | Breaks subsequent tests that need the value | Save and restore in guaranteed cleanup |
+| In-test retry loops for flaky tests | Masks root cause, inflates suite duration | Root-cause and fix or quarantine |
+| Test interdependence | Test B depends on state from Test A; reorder breaks both | Each test sets up and tears down its own state |
+| Assertion-free tests | Test runs code but never asserts; passes as long as nothing throws | Every test must assert on a specific expected outcome |
+| Sleep-based synchronization | `sleep(2)` to wait for async work; slow and still flaky | Use polling with timeout, await, or synchronization primitives |
+| Hardcoded colliding test data | Every test uses `id=1` or `email=test@example.com`; parallel runs collide | Generate unique identifiers per test |
+| Over-mocking | So many mocks that the test encodes the implementation, not behavior; brittle to refactors | Mock only direct dependencies; let integration tests cover wiring |
+| Mocking third-party internals | Mocking a library's internal API rather than its public interface | Use the library's test utilities or mock at your own abstraction boundary |
```
Comment (jcdendrite) [inline claude/.claude/skills/test-conventions/SKILL.md]: Same comment as above. Better for test evaluation, like a code review skill or reference specific for test code.

File: claude/.claude/settings.json [hunk 10]
```diff
@@ -18,5 +17,8 @@
     "type": "command",
     "command": "bash ~/.claude/statusline-command.sh"
   },
-  "effortLevel": "high"
+  "effortLevel": "high",
+  "enabledPlugins": {
```
Comment (jcdendrite) [inline claude/.claude/settings.json]: Don't commit this part about enabledPlugins. Can you put it in settings.local.json or does it need to be installed at the user level because it applies to about 7-10 repos for one of the projects I'm working on?

**Review by jcdendrite — COMMENTED** [review]
Comments for Claude Code

File: claude/.claude/skills/test-conventions/SKILL.md (hunk 1 above)
Comment (jcdendrite) [reply claude/.claude/skills/test-conventions/SKILL.md]: **[Claude Code]** Agree — the inverted pyramid diagnostics are more useful when evaluating an existing suite than when writing new tests. Will move this and the related evaluation sections (flaky test diagnosis, anti-patterns) to a separate test-evaluation skill, keeping this skill focused on writing conventions.

File: claude/.claude/skills/test-conventions/SKILL.md (hunk 2 above)
Comment (jcdendrite) [reply claude/.claude/skills/test-conventions/SKILL.md]: **[Claude Code]** You're right — these are heuristics, not rules. Will change the heading to "When to consider moving a test down the pyramid" and soften the language to signal guidance rather than mandates.

File: claude/.claude/skills/test-conventions/SKILL.md (hunk 3 above)
Comment (jcdendrite) [reply claude/.claude/skills/test-conventions/SKILL.md]: **[Claude Code]** Agreed — "relative to now" is a flaky test source (midnight crossings, TZ differences, DST transitions). Will remove it as a recommendation and keep only "inject timestamps as parameters."

File: claude/.claude/skills/test-conventions/SKILL.md (hunk 4 above)
Comment (jcdendrite) [reply claude/.claude/skills/test-conventions/SKILL.md]: **[Claude Code]** Good point — the section jumps straight to managing global state without first saying to avoid it. Will add a preamble: prefer designs that don't require global state in tests; when unavoidable, save and restore.

File: claude/.claude/skills/test-conventions/SKILL.md (hunk 5 above)
Comment (jcdendrite) [reply claude/.claude/skills/test-conventions/SKILL.md]: **[Claude Code]** Fair — the parallel-safe guidance applies equally to local development. Will rephrase the intro to cover both local and CI contexts.

File: claude/.claude/skills/test-conventions/SKILL.md (hunk 6 above)
Comment (jcdendrite) [reply claude/.claude/skills/test-conventions/SKILL.md]: **[Claude Code]** Good catch. The distinction: integration tests verify your code'\''s actual response shape (the real output from your endpoint). Contract tests verify schema compatibility between services without calling the other service. Will add a clarifying sentence to draw that boundary.

File: claude/.claude/skills/test-conventions/SKILL.md (hunk 7 above)
Comment (jcdendrite) [reply claude/.claude/skills/test-conventions/SKILL.md]: **[Claude Code]** Agreed — smoke tests and E2E tests serve different purposes. Smoke tests are lightweight post-deploy health checks ("is the critical path up?"). E2E tests verify complete user flows end-to-end. Will split them into separate pyramid rows.

File: claude/.claude/skills/test-conventions/SKILL.md (hunk 8 above)
Comment (jcdendrite) [reply claude/.claude/skills/test-conventions/SKILL.md]: **[Claude Code]** Agree — will move the flaky test diagnosis list to the evaluation skill. Though a few preventive items ("don'\''t use sleep-based sync", "don'\''t hardcode colliding data") are worth keeping in the writing skill since they prevent mistakes during authoring rather than diagnosing them after.

File: claude/.claude/skills/test-conventions/SKILL.md (hunk 9 above)
Comment (jcdendrite) [reply claude/.claude/skills/test-conventions/SKILL.md]: **[Claude Code]** Same plan — will move the full anti-patterns table to the evaluation skill. Will keep a short list of the most common authoring-time mistakes (assertion-free tests, sleep-based sync, hardcoded colliding data, over-mocking) inline in the writing skill.

File: claude/.claude/settings.json (hunk 10 above)
Comment (jcdendrite) [reply claude/.claude/settings.json]: **[Claude Code]** This file shouldn'\''t be in the PR — the branch inherited 2 prior commits from main that were ahead of origin/main. Will clean up the PR scope.

For the plugin question: `enabledPlugins` works at all config levels (user, project, local). Since this is a dotfiles repo that deploys to `~/.claude/`, the cleanest option is to keep `enabledPlugins` in `~/.claude/settings.json` but exclude it from the committed dotfiles version. You could either:
1. Add it to a `settings.local.json` overlay that gets applied at deploy time but isn'\''t committed
2. Configure it directly in `~/.claude/settings.json` on each machine after dotfiles deploy

Option 1 is cleaner for a dotfiles workflow. Either way, the plugin will be available across all repos without committing it to the public repo.

Comment (jcdendrite) [reply claude/.claude/settings.json]: Let's go with Option #1. settings.local.json shouldn't be committed anyways

File: claude/.claude/skills/test-conventions/SKILL.md [hunk 11]
```diff
@@ -90,7 +75,7 @@ When replacing globals (HTTP client, fetch function, clock), always restore the
 - **In-memory vs. real engine:** In-memory databases (e.g., SQLite in-memory mode) are fast but have dialect differences (JSON columns, CTEs, locking behavior). When dialect fidelity matters, use the same engine as production.
 
 ### Parallel-safe tests
-When tests run in parallel (pytest-xdist, Jest workers, Go's `t.Parallel()`, JUnit parallel mode):
+Whether running locally or in CI, parallel test runners (pytest-xdist, Jest workers, Go's `t.Parallel()`, JUnit parallel mode) introduce isolation requirements:
 - Never bind to hardcoded ports; use port 0 or dynamic allocation
```
Comment (jcdendrite) [inline claude/.claude/skills/test-conventions/SKILL.md]: Sorry, I meant this instruction about binding to hardcoded ports might be different for local setups versus CI...

File: claude/.claude/skills/test-conventions/SKILL.md [hunk 12]
```diff
@@ -20,31 +21,12 @@ user-invocable: false
...
-- The test sets up complex DB state to exercise one `if` branch
-- The test hits an external API just to test error handling
-- The test is slow (>500ms) and tests pure logic, not wiring
-- You're hitting rate limits on external APIs
-
-**Smoke → integration when:**
-- The test calls a real third-party API to verify your code handles errors
-- The test is flaky because of network conditions
-- The test creates real resources (emails, contacts) as a side effect
+**E2E tests** verify complete user journeys across the full stack. **Smoke tests** are lightweight post-deploy checks that verify critical paths are functional — they answer "is it up?" not "does every flow work?"
```
Comment (jcdendrite) [inline claude/.claude/skills/test-conventions/SKILL.md]: Formatting inconsistency. These got pulled onto the same line, and there is a newline in between E2E and contract tests. I'm committing a fix.

File: claude/.claude/skills/test-conventions/SKILL.md [hunk 13]
```diff
@@ -135,20 +120,24 @@ For tests that guard against a specific past bug, include a comment or docstring
...
-- **HTTP-level:** Auth rejection (401/403), authorized success (2xx), response shape (expected fields and types)
+- **HTTP-level:** Auth rejection (401/403), authorized success (2xx), response shape — the actual output your endpoint produces (fields, types, status codes)
 - **Service/module-level:** Wiring between layers maps results correctly, shared modules are called with expected arguments
 - Not every integration test needs to go through HTTP — service-level integration tests are cheaper when you're verifying wiring, not auth or response shape
 - Don't re-test every branch — that's the unit tests' job
 
 **Contract tests:**
-- Consumer expectations match the provider's actual schema
+- Consumer expectations match the provider's actual API schema — the agreed-upon interface between services, independent of either side's implementation
+- Distinct from integration response shape tests: contract tests verify the *schema agreement* between services; integration tests verify your *code's actual output*
```
Comment (jcdendrite) [inline claude/.claude/skills/test-conventions/SKILL.md:130]: You may want to give more guidance as to when to make a response shape-based Integration test rather than a response shape-based contract test.

File: claude/.claude/skills/test-evaluation/SKILL.md [hunk 14]
```diff
@@ -0,0 +1,76 @@
...
+
+These are heuristics, not rules — evaluate each case against the specific codebase's constraints.
+
+**Integration → unit when:**
+- The test sets up complex DB state to exercise one `if` branch
+- The test hits an external API just to test error handling
+- The test is slow (>500ms) and tests pure logic, not wiring
+- You're hitting rate limits on external APIs
+
+**E2E/Smoke → integration when:**
```
Comment (jcdendrite) [inline claude/.claude/skills/test-evaluation/SKILL.md]: E2E - > integration makes sense. Smoke tests have a totally different purpose - I don't think you can provide guidance to convert a smoke test to an integration test...

File: claude/.claude/settings.json [hunk 15]
```diff
@@ -1,4 +1,5 @@
 {
+  "model": "Opus",
```
Comment (jcdendrite) [inline claude/.claude/settings.json]: let's not commit the model

File: claude/.claude/settings.json [hunk 16]
```diff
@@ -17,8 +18,5 @@
     "type": "command",
     "command": "bash ~/.claude/statusline-command.sh"
   },
-  "effortLevel": "high",
-  "enabledPlugins": {
-    "atlassian@claude-plugins-official": true
-  }
+  "effortLevel": "high"
```
Comment (jcdendrite) [inline claude/.claude/settings.json]: also shouldn't commit effort level

File: claude/.claude/skills/test-conventions/SKILL.md (hunk 11 above)
Comment (jared-yetimyfriend) [reply claude/.claude/skills/test-conventions/SKILL.md]: **[Claude Code]** Good point — locally you might intentionally use a fixed port for convenience (e.g., localhost:3000 for a dev server). The "never hardcode" advice is specifically about tests that may run in parallel. Will qualify this to distinguish test port binding from dev server conventions.

File: claude/.claude/skills/test-conventions/SKILL.md (hunk 12 above)
Comment (jared-yetimyfriend) [reply claude/.claude/skills/test-conventions/SKILL.md]: **[Claude Code]** Noted — your formatting fix is on disk, will include it in the next commit.

File: claude/.claude/skills/test-conventions/SKILL.md (hunk 13 above)
Comment (jared-yetimyfriend) [reply claude/.claude/skills/test-conventions/SKILL.md:130]: **[Claude Code]** Will add decision criteria. The key distinction: use integration response shape tests when you own both sides (or there'\''s only one consumer) and can test against a running instance. Use contract tests when provider and consumer are owned by different teams, deploy independently, or when spinning up the provider in tests is expensive.

File: claude/.claude/skills/test-evaluation/SKILL.md (hunk 14 above)
Comment (jared-yetimyfriend) [reply claude/.claude/skills/test-evaluation/SKILL.md]: **[Claude Code]** You'\''re right — smoke tests verify deployment health, not behavior. You wouldn'\''t "move a smoke test down" because it'\''s not testing the same thing as integration tests. Will change this to only cover E2E → integration.

File: claude/.claude/settings.json (hunk 15 above)
Comment (jared-yetimyfriend) [reply claude/.claude/settings.json]: **[Claude Code]** Will move both model and effortLevel to settings.local.json — these are personal preferences, not shared config.

File: claude/.claude/settings.json (hunk 16 above)
Comment (jared-yetimyfriend) [reply claude/.claude/settings.json]: **[Claude Code]** Agreed — moving to settings.local.json with model.
